corehq/util/doc_processor/interface.py

```diff
@@ -0,0 +1,314 @@
+import weakref
+from abc import ABCMeta, abstractmethod
+from datetime import datetime
+
+import six
+
+from corehq.util.pagination import PaginationEventHandler, TooManyRetries
+
+
+class BulkProcessingFailed(Exception):
+    pass
+
+
+DOCS_SKIPPED_WARNING = """
+        WARNING {} documents were not processed due to concurrent modification
+        during migration. Run the migration again until you do not see this
+        message.
+        """
+
+
+class BaseDocProcessor(six.with_metaclass(ABCMeta)):
+    """Base class for processors that get passed"""
+
+    def __enter__(self):
+        pass
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        pass
+
+    def process_doc(self, doc):
+        """Process a single document
+
+        :param doc: The document dict to be processed.
+        :returns: True if doc was processed successfully else False. If this returns False
+        the document migration will be retried later.
+        """
+        raise NotImplementedError
+
+    def process_bulk_docs(self, docs):
+        """Process a batch of documents. The default implementation passes
+        each doc in turn to ``process_doc``.
+
+        :param docs: A list of document dicts to be processed.
+        :returns: True if doc was processed successfully else False.
+        If this returns False the processing will be halted.
+        """
+        return all(self.process_doc(doc) for doc in docs)
+
+    def handle_skip(self, doc):
+        """Called when a document is going to be skipped i.e. it has been
+        retried > max_retries.
+
+        :returns: True to indicate that the skip has been handled
+                  or False to stop execution
+        """
+        return False
+
+    def processing_complete(self, skipped):
+        pass
+
+    def should_process(self, doc):
+        """
+        :param doc: the document to filter
+        :return: True if this doc should be migrated
+        """
+        return True
+
+
+class ProcessorProgressLogger(object):
+    def progress_starting(self, total, previously_visited):
+        print("Processing {} documents{}: ...".format(
+            total,
+            " (~{} already processed)".format(previously_visited) if previously_visited else ""
+        ))
+
+    def document_skipped(self, doc_dict):
+        print("Skip: {doc_type} {_id}".format(**doc_dict))
+
+    def progress(self, total, processed, visited, time_elapsed, time_remaining):
+        print("Processed {}/{} of {} documents in {} ({} remaining)"
+              .format(processed, visited, total, time_elapsed, time_remaining))
+
+    def progress_complete(self, total, processed, visited, previously_visited, filtered):
```

Comment:
args order does not match [the call](https://github.com/dimagi/commcare-hq/pull/12326/files#diff-da7fddbada9ec9a3b6760d0f91093b4eR236). I'm seeing output like:
```
Processed 7055/3088965 of 600 documents in 0:00:45.943778 (5:34:29.397920 remaining)
```